src/routers/Gateway.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.18;
+pragma abicoder v2;
+
+import {TypedMemView} from "memview-sol/TypedMemView.sol";
+import {ConnectorMessages} from "../Messages.sol";
+
+interface ConnectorLike {
+    function addPool(uint64 poolId) external;
+    function addTranche(
+        uint64 poolId,
+        bytes16 trancheId,
+        string memory tokenName,
+        string memory tokenSymbol,
+        uint128 price
+    ) external;
+    function updateMember(uint64 poolId, bytes16 trancheId, address user, uint64 validUntil) external;
+    function updateTokenPrice(uint64 poolId, bytes16 trancheId, uint128 price) external;
+    function handleTransfer(uint64 poolId, bytes16 trancheId, address destinationAddress, uint128 amount) external;
+}
+
+interface RouterLike {
+    function send(bytes memory message) external;
+}
+
+contract ConnectorGateway {
+    using TypedMemView for bytes;
+    // why bytes29? - https://github.com/summa-tx/memview-sol#why-bytes29
+    using TypedMemView for bytes29;
+    using ConnectorMessages for bytes29;
+
+    mapping(address => uint256) public wards;
+
+    ConnectorLike public immutable connector;
+    // TODO: support multiple incoming routers (just a single outgoing router) to simplify router migrations
+    RouterLike public immutable router;
+
+    /// --- Events ---
+    event Rely(address indexed user);
+    event Deny(address indexed user);
+    event File(bytes32 indexed what, address addr);
+
+    constructor(address connector_, address router_) {
+        connector = ConnectorLike(connector_);
+        router = RouterLike(router_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    modifier auth() {
+        require(wards[msg.sender] == 1, "ConnectorGateway/not-authorized");
+        _;
+    }
+
+    modifier onlyConnector() {
+        require(msg.sender == address(connector), "ConnectorGateway/only-connector-allowed-to-call");
+        _;
+    }
+
+    modifier onlyRouter() {
+        require(msg.sender == address(router), "ConnectorGateway/only-router-allowed-to-call");
+        _;
+    }
+
+    // --- Administration ---
+    function rely(address user) external auth {
+        wards[user] = 1;
+        emit Rely(user);
+    }
+
+    function deny(address user) external auth {
+        wards[user] = 0;
+        emit Deny(user);
+    }
+
+    // --- Outgoing ---
+    function transfer(
+        uint64 poolId,
+        bytes16 trancheId,
+        ConnectorMessages.Domain destinationDomain,
+        bytes32 destinationAddress,
+        uint128 amount
+    ) public onlyConnector {
+        router.send(
```

Comment:
@offerijns @AStox this sort of presumes that the single router instance will know how to router messages to any domain, whereas our current setup is to have a XCM router, which wouldn't know how to route a message to an EVM chain. How will this work in the future?

Reply:
Oh ok, that's new to me 👍